contracts/non-fungible-tokens/andromeda-crowdfund/src/contract.rs

```diff
@@ -534,6 +538,99 @@ fn transfer_asset_msg(
+
+    let sub_response = match curr_stage {
+        CampaignStage::SUCCESS => handle_successful_claim(deps.branch(), &info.sender)?,
+        CampaignStage::FAILED => handle_failed_claim(deps.branch(), &info.sender)?,
+        _ => {
+            return Err(ContractError::InvalidCampaignOperation {
+                operation: "Claim".to_string(),
+                stage: curr_stage.to_string(),
+            })
+        }
+    };
+    resp = resp
+        .add_attributes(sub_response.attributes)
+        .add_submessages(sub_response.messages);
+
+    clear_user_orders(deps.storage, info.sender)?;
+
+    Ok(resp)
+}
+
+fn handle_successful_claim(deps: DepsMut, sender: &Addr) -> Result<Response, ContractError> {
+    let campaign_config = get_config(deps.storage)?;
+
+    let orders = get_user_orders(deps.storage, sender.clone(), None, None, true)?;
+    ensure!(!orders.is_empty(), ContractError::NoPurchases {});
+
+    // mint tier token to the owner
+    let tier_address = campaign_config
+        .tier_address
+        .get_raw_address(&deps.as_ref())?;
+
+    let mut resp = Response::new();
+    for order in orders {
+        let meta_data = get_tier(deps.storage, order.level.into())?.meta_data;
+        for _ in 0..order.amount.into() {
+            let mint_resp = mint(
+                deps.storage,
+                tier_address.to_string(),
+                meta_data.clone(),
+                sender.to_string(),
+            )?;
+            resp = resp
+                .add_attributes(mint_resp.attributes)
+                .add_submessages(mint_resp.messages);
+        }
+    }
+    Ok(resp)
+}
+
+fn handle_failed_claim(deps: DepsMut, sender: &Addr) -> Result<Response, ContractError> {
+    let campaign_config = get_config(deps.storage)?;
+
+    let orders = get_user_orders(deps.storage, sender.clone(), None, None, false)?;
+    ensure!(!orders.is_empty(), ContractError::NoPurchases {});
+
+    // refund
+    let total_cost = orders.iter().try_fold(Uint128::zero(), |sum, order| {
+        let tier = get_tier(deps.storage, u64::from(order.level))?;
+        let new_sum: Result<Uint128, ContractError> = Ok(sum + tier.price * order.amount);
+        new_sum
+    })?;
+    let mut resp = Response::new();
+    let sub_msg = transfer_asset_msg(sender.to_string(), total_cost, campaign_config.denom)?;
+    resp = resp.add_submessage(sub_msg);
+
+    Ok(resp)
+}
+
+fn mint(
+    storage: &mut dyn Storage,
+    tier_contract: String,
+    tier_meta_data: TierMetaData,
+    owner: String,
+) -> Result<Response, ContractError> {
+    let token_id = get_and_increase_tier_token_id(storage)?.to_string();
+
+    Ok(Response::new().add_message(WasmMsg::Execute {
+        contract_addr: tier_contract,
+        msg: encode_binary(&Cw721ExecuteMsg::Mint {
+            token_id,
+            owner,
+            token_uri: tier_meta_data.token_uri,
+            extension: tier_meta_data.extension,
```

Comment:
We need to ensure that this extension type matches the contract its minting through.

Reply:
I am not sure how to ensure this. extension is defined as follow.
```
pub struct TokenExtension {
    /// The original publisher of the token
    pub publisher: String,
}
```
Do we need to ensure that `publisher` remains constant for all tiers?